File: src/main/java/com/wolt/domain/deliveryfeecalculator/DeliveryDistanceCalculator.java
package com.wolt.domain.deliveryfeecalculator;

import java.math.BigInteger;

class DeliveryDistanceCalculator {
    private final static BigInteger BASE_DISTANCE = BigInteger.valueOf(500);
    private final static BigInteger BASE_FEE_PER_500_METERS = BigInteger.valueOf(1_00);

    BigInteger calculate(BigInteger deliveryDistance) {
        BigInteger baseFee = deliveryDistance.mod(BASE_DISTANCE)
                .signum() > 0
                ? BASE_FEE_PER_500_METERS
                : BigInteger.ZERO;
        BigInteger multiplicationFee = deliveryDistance.divide(BASE_DISTANCE)
                .multiply(BASE_FEE_PER_500_METERS);
        return baseFee.add(multiplicationFee);
    }
}

File: src/main/java/com/wolt/domain/deliveryfeecalculator/DeliveryDistanceCalculator.kt
package com.wolt.domain.deliveryfeecalculator

import java.math.BigInteger

internal class DeliveryDistanceCalculator {
    fun calculate(deliveryDistance: BigInteger): BigInteger {
        if (isBaseDistance(deliveryDistance)) {
            return BASE_FEE_PER_500_METERS
        }
        val deliveryFee = deliveryDistance.divide(BASE_DISTANCE)
            .multiply(BASE_FEE_PER_500_METERS)
        return if (isTotalDistanceIncluded(deliveryDistance)) {
            deliveryFee
        } else deliveryFee.add(BASE_FEE_PER_500_METERS)
    }

    private fun isBaseDistance(deliveryDistance: BigInteger): Boolean {
        return deliveryDistance <= BASE_DISTANCE
    }

    private fun isTotalDistanceIncluded(deliveryDistance: BigInteger): Boolean {
        return deliveryDistance.remainder(BASE_FEE_PER_500_METERS) <= BigInteger.ZERO
    }

    companion object {
        private val BASE_DISTANCE = BigInteger.valueOf(500)
        private val BASE_FEE_PER_500_METERS = BigInteger.valueOf(1_00)
    }
}